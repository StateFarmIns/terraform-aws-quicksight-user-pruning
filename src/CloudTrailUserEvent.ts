import { Event } from '@aws-sdk/client-cloudtrail'

export class CloudTrailUserEvent {
	iamRoleId: string // ASIAGREHGHTHRTH3XJG12
	email: string
	eventTime: Date

	constructor(event: Event) {
		const cloudTrailEvent = JSON.parse(event.CloudTrailEvent)
		const [iamRoleId, email] = cloudTrailEvent.userIdentity.principalId.split(':')

		this.iamRoleId = iamRoleId
		this.email = email
		this.eventTime = event.EventTime
	}
}