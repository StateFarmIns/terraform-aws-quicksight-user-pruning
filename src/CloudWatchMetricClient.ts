import { CloudWatchClient, MetricDatum, PutMetricDataCommand } from '@aws-sdk/client-cloudwatch'

export class CloudWatchMetricClient {
	private metricData: MetricDatum[] = []
	private cloudWatchClient = new CloudWatchClient({})

	// Normally we'd worry about max metric data per request but the max is currently 20 and we won't be publishing that many metrics
	public queueMetric(metricDatum: MetricDatum) {
		this.metricData.push(metricDatum)
	}

	public async emitQueuedMetrics() {
		const putMetricDataCommand = new PutMetricDataCommand({
			Namespace: process.env.AWS_LAMBDA_FUNCTION_NAME,
			MetricData: this.metricData,
		})

		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`Emitting metrics: ${JSON.stringify(putMetricDataCommand)}`)

		await this.cloudWatchClient.send(putMetricDataCommand)

		this.metricData = []
	}
}