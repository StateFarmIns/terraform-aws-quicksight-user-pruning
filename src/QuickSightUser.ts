import { User } from '@aws-sdk/client-quicksight'

export class QuickSightUser {
	arn: string // arn:aws:quicksight:<REGION>:<ACCOUNT>:user/default/<IAM ROLE NAME>/<STS SESSION>
	email: string
	lastAccess: Date // Comes from CloudTrail
	role: QuickSightRole
	username: string // <IAM ROLE>/<STS SESSION>
	iamRole: string
	stsSession: string
	invalid: boolean // User is invalid if username is "N/A". See README.md for more details

	constructor(quickSightUser: User) {
		this.arn = quickSightUser.Arn
		this.email = quickSightUser.Email
		this.role = QuickSightRole[quickSightUser.Role as keyof typeof QuickSightRole] // Only values that will be returned // also the keyof typeof hack came from here https://stackoverflow.com/a/42623905
		this.username = quickSightUser.UserName
		this.invalid = this.username === 'N/A'
		const [iamRole, stsSession] = this.username.split('/')
		this.iamRole = iamRole
		this.stsSession = stsSession
	}
}

export enum QuickSightRole {
	READER = 'READER', AUTHOR = 'AUTHOR', ADMIN = 'ADMIN'
}