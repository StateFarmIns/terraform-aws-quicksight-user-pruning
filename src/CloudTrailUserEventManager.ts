import { CloudTrailClient, LookupEventsCommand } from '@aws-sdk/client-cloudtrail'
import { CloudTrailUserEvent } from './CloudTrailUserEvent'
import { QuickSightUser } from './QuickSightUser'

export class CloudTrailUserEventManager {
	private cloudTrailClient = new CloudTrailClient({})

	// Filters out events that do not originate with someone's e-mail address
	// such as any automated process which touches QuickSight, including this one
	public async retrieveQuickSightUserEvents(startDate: Date): Promise<CloudTrailUserEvent[]> {
		const events: CloudTrailUserEvent[] = []

		let nextToken: string = null // This is annoying; null doesn't work in the QuickSight API but empty string doesn't work in the CloudTrail API
		do {
			const lookupEventsCommand = new LookupEventsCommand({
				LookupAttributes: [
					{
						AttributeKey: 'EventSource',
						AttributeValue: 'quicksight.amazonaws.com',
					},
				],
				StartTime: startDate,
				NextToken: nextToken,
			})

			const lookupEventsResult = await this.cloudTrailClient.send(lookupEventsCommand)

			nextToken = lookupEventsResult.NextToken

			lookupEventsResult.Events.forEach((event) => events.push(new CloudTrailUserEvent(event)))
		} while (nextToken)

		return events
	}

	public getLastAccessDate(user: QuickSightUser, events: CloudTrailUserEvent[]): Date {
		const thisParticularUserEvents = events.filter((event) => user.iamRole === event.iamRole && user.stsSession === event.stsSession)

		if (thisParticularUserEvents.length === 0) {
			// Stryker disable next-line all "I do not care about mutating console statements"
			console.debug(`${user.iamRole}/${user.stsSession} no CloudTrail events found`)
			return new Date(0) // Return 1970 as last access date to make logic easier down the road
		}

		// Sort newest to oldest
		const newestEvent = thisParticularUserEvents.sort((a, b) => a.eventTime <= b.eventTime ? 1 : -1)[0]
		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`${user.iamRole}/${user.stsSession} latest event: ${JSON.stringify(newestEvent, null, 1)}`)

		return newestEvent.eventTime
	}
}