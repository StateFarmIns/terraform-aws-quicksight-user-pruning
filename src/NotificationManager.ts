import { SendEmailCommand, SESClient } from '@aws-sdk/client-ses'
import { QuickSightUser } from './QuickSightUser'

export class NotificationManager {
	private sesClient = new SESClient({})
	private noLastAccessDate = new Date(0).toLocaleString()

	public async notifyUser(user: QuickSightUser) {
		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`Notifying ${JSON.stringify(user, null, 1)}`)
		
		// If the date is 1970, we didn't find any activity for the user
		const lastAccess =
			user.lastAccess.toLocaleString() === this.noLastAccessDate
				? `not in the last ${process.env.deleteDays} days`
				: user.lastAccess.toLocaleDateString()

		const subject = `I: QuickSight user ${user.username} in account ${process.env.accountAlias} will be deleted`
		const body = `In order to <a href="https://aws.amazon.com/quicksight/pricing/">control costs ($24/month/user)</a>,
    your AWS QuickSight user '${user.username}' 
    will be deleted when you have not accessed QuickSight in ${process.env.deleteDays} days. 
    CloudTrail records indicate that your last access was ${lastAccess}.<br /><br />
    
    You can keep your user by logging in and taking actions in QuickSight. 
    Alternatively, you can do nothing, and your user will be automatically deleted. 
    Never fear, you simply log back into QuickSight and enter your e-mail, and your user gets recreated. 
    No dashboards or other resources will be deleted at this time (we may clean those up in the future, with notification, of course).<br /><br />
    
    Account Number: ${process.env.awsAccountId}<br />
    Account Alias: ${process.env.accountAlias}<br />
    Assumed Role: ${user.iamRole}<br />
    Email: ${user.email}<br /><br />
  
    For further assistance, contact <a href="mailto:${process.env.contact}">${process.env.contact}</a>. Thank you and have a fantastically fun-filled day ðŸ¥³.
    `

		const sendEmailCommand = new SendEmailCommand({
			Source: `${process.env.accountAlias} Cloud Account Administrator<${process.env.from}>`,
			Destination: {
				ToAddresses: [user.email],
				CcAddresses: JSON.parse(process.env.cc),
			},
			ReplyToAddresses: [process.env.replyTo],
			SourceArn: process.env.sesArn,
			Message: {
				Subject: { Data: subject },
				Body: {
					Html: { Data: body },
				},
			},
		})

		await this.sesClient.send(sendEmailCommand)
	}
}