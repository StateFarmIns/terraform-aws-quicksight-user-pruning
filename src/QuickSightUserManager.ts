import { DeleteUserCommand, ListUsersCommand, QuickSightClient } from '@aws-sdk/client-quicksight'
import { QuickSightUser } from './QuickSightUser'

export class QuickSightUserManager {
	private awsAccountId = process.env.awsAccountId
	private static namespace = 'default'
	private quickSightClient = new QuickSightClient({})
  
	public async retrieveUsers(): Promise<QuickSightUser[]> {
		const quickSightUsers: QuickSightUser[] = []

		let nextToken = '' // originally I tried null, but it causes API signature issues.
		do {
			const listUsersCommand = new ListUsersCommand({ AwsAccountId: this.awsAccountId, Namespace: QuickSightUserManager.namespace, NextToken: nextToken })
			const quickSightUserListResponse = await this.quickSightClient.send(listUsersCommand)

			nextToken = quickSightUserListResponse.NextToken

			quickSightUserListResponse.UserList.forEach((user) => quickSightUsers.push(new QuickSightUser(user)))
		} while (nextToken)

		return quickSightUsers
	}

	public async deleteUser(user: QuickSightUser) {
		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`Deleting ${JSON.stringify(user, null, 1)}`)
		
		const deleteUserCommand = new DeleteUserCommand({
			AwsAccountId: this.awsAccountId,
			Namespace: QuickSightUserManager.namespace,
			UserName: user.username,
		})

		await this.quickSightClient.send(deleteUserCommand)
	}
}