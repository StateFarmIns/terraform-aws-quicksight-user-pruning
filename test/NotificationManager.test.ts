import { SendEmailCommand, SESClient } from '@aws-sdk/client-ses'
import { mockClient } from 'aws-sdk-client-mock'
import { NotificationManager } from '../src/NotificationManager'
import { QuickSightRole, QuickSightUser } from '../src/QuickSightUser'

process.env = {
	...process.env,
	awsAccountId: '1234567890',
	accountAlias: 'MyAccountAlias',
	contact: 'contact@example.com',
	from: 'from@example.com',
	cc: JSON.stringify(['cc1@example.com', 'cc2@example.com']),
	replyTo: 'replyTo@example.com',
	sesArn: 'arn:aws:ses:us-east-1:1234567890:identity/example.com',
	deleteDays: '30',
}

const notificationManager = new NotificationManager()
const sesClientMock = mockClient(SESClient)

const validQuickSightUser: QuickSightUser = {
	arn: 'arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-admin-role/john.smith@example.com',
	email: 'john.smith@example.com',
	role: QuickSightRole.ADMIN,
	username: 'quicksight-admin-role/john.smith@example.com',
	iamRole: 'quicksight-admin-role',
	stsSession: 'john.smith@example.com',
	invalid: false,
	lastAccess: new Date('2022-05-09T01:02:03'),
}

const userNotAccessedRecently = { ...validQuickSightUser, lastAccess: new Date(0) }

afterEach(() => {
	sesClientMock.reset()
})

describe('NotificationManager', () => {
	describe('successful when', () => {
		it('gets a valid user', async () => {
			sesClientMock.on(SendEmailCommand).resolves({})

			await notificationManager.notifyUser(validQuickSightUser)

			expect(sesClientMock.calls()).toHaveLength(1)
			expect(sesClientMock.call(0).args[0].input).toMatchSnapshot()
		})

		it('gets a user who has not accessed QS recently', async () => {
			sesClientMock.on(SendEmailCommand).resolves({})

			await notificationManager.notifyUser(userNotAccessedRecently)

			expect(sesClientMock.calls()).toHaveLength(1)
			expect(sesClientMock.call(0).args[0].input).toMatchSnapshot()
		})
	})

	describe('fails when', () => {
		// Intentionally we don't handle errors in the Lambda, so just checking to make sure that it bubbles up the exception
		it('receives a SES client error', async () => {
			const error: Error = { name: 'MyError', message: 'Oh no!' }
			sesClientMock.on(SendEmailCommand).rejects(error)

			await expect(notificationManager.notifyUser(validQuickSightUser)).rejects.toThrowError(error)
		})
	})
})