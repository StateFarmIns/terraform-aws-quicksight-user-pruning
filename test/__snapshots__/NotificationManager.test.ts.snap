// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationManager successful when gets a user who has not accessed QS recently 1`] = `
Object {
  "Destination": Object {
    "CcAddresses": Array [
      "cc1@example.com",
      "cc2@example.com",
    ],
    "ToAddresses": Array [
      "john.smith@example.com",
    ],
  },
  "Message": Object {
    "Body": Object {
      "Html": Object {
        "Data": "In order to <a href=\\"https://aws.amazon.com/quicksight/pricing/\\">control costs ($24/month/user)</a>,
    your AWS QuickSight user 'quicksight-admin-role/john.smith@example.com' 
    will be deleted when you have not accessed QuickSight in 30 days. 
    CloudTrail records indicate that your last access was not in the last 30 days.<br /><br />
    
    You can keep your user by logging in and taking actions in QuickSight. 
    Alternatively, you can do nothing, and your user will be automatically deleted. 
    Never fear, you simply log back into QuickSight and enter your e-mail, and your user gets recreated. 
    No dashboards or other resources will be deleted at this time (we may clean those up in the future, with notification, of course).<br /><br />
    
    Account Number: 1234567890<br />
    Account Alias: MyAccountAlias<br />
    Assumed Role: quicksight-admin-role<br />
    Email: john.smith@example.com<br /><br />
  
    For further assistance, contact <a href=\\"mailto:contact@example.com\\">contact@example.com</a>. Thank you and have a fantastically fun-filled day ðŸ¥³.
    ",
      },
    },
    "Subject": Object {
      "Data": "I: QuickSight user quicksight-admin-role/john.smith@example.com in account MyAccountAlias will be deleted",
    },
  },
  "ReplyToAddresses": Array [
    "replyTo@example.com",
  ],
  "Source": "MyAccountAlias Cloud Account Administrator<from@example.com>",
  "SourceArn": "arn:aws:ses:us-east-1:1234567890:identity/example.com",
}
`;

exports[`NotificationManager successful when gets a valid user 1`] = `
Object {
  "Destination": Object {
    "CcAddresses": Array [
      "cc1@example.com",
      "cc2@example.com",
    ],
    "ToAddresses": Array [
      "john.smith@example.com",
    ],
  },
  "Message": Object {
    "Body": Object {
      "Html": Object {
        "Data": "In order to <a href=\\"https://aws.amazon.com/quicksight/pricing/\\">control costs ($24/month/user)</a>,
    your AWS QuickSight user 'quicksight-admin-role/john.smith@example.com' 
    will be deleted when you have not accessed QuickSight in 30 days. 
    CloudTrail records indicate that your last access was 5/9/2022.<br /><br />
    
    You can keep your user by logging in and taking actions in QuickSight. 
    Alternatively, you can do nothing, and your user will be automatically deleted. 
    Never fear, you simply log back into QuickSight and enter your e-mail, and your user gets recreated. 
    No dashboards or other resources will be deleted at this time (we may clean those up in the future, with notification, of course).<br /><br />
    
    Account Number: 1234567890<br />
    Account Alias: MyAccountAlias<br />
    Assumed Role: quicksight-admin-role<br />
    Email: john.smith@example.com<br /><br />
  
    For further assistance, contact <a href=\\"mailto:contact@example.com\\">contact@example.com</a>. Thank you and have a fantastically fun-filled day ðŸ¥³.
    ",
      },
    },
    "Subject": Object {
      "Data": "I: QuickSight user quicksight-admin-role/john.smith@example.com in account MyAccountAlias will be deleted",
    },
  },
  "ReplyToAddresses": Array [
    "replyTo@example.com",
  ],
  "Source": "MyAccountAlias Cloud Account Administrator<from@example.com>",
  "SourceArn": "arn:aws:ses:us-east-1:1234567890:identity/example.com",
}
`;
